File: go/escape.go
/* ********************************************************************** */
/**
 * @brief   Command line tool to escape characters greater than 0x7F.
 * @author  eel3
 * @date    2013/01/22
 *
 * @par Compilers
 * - Go 1.3.0 windows/amd64
 */
/* ********************************************************************** */

package main

import (
	"flag"
	"fmt"
	"io"
	"os"
	"path"
)

const (
	exit_success = iota // EXIT_SUCCESS
	exit_failure        // EXIT_FAILURE
)

// Initialize function (specify help message).
func init() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "usage: %s [-o out-file] [file ...]\n", path.Base(os.Args[0]))
		flag.PrintDefaults()
	}
}

// Escape characters greater than 0x7F.
func doEscape(in *os.File, out *os.File) {
	buf := make([]byte, 1)
	for {
		n, err := in.Read(buf)
		if n == 0 && err == io.EOF {
			break
		}
		if buf[0] >= 0x80 {
			fmt.Fprintf(out, `\x%X`, buf[0])
		} else {
			out.Write(buf)
		}
	}
}

// Read file at `infile' and escape all text.
func escapeFile(infile string, out *os.File) error {
	var in *os.File

	if infile == "-" {
		in = os.Stdin
	} else {
		var err error
		in, err = os.Open(infile)
		if err != nil {
			return err
		}
		defer in.Close()
	}

	doEscape(in, out)
	return nil
}

// Main routine.
func main() {
	outfile := flag.String("o", "-", "output file or `-' (stdout)")
	flag.Parse()

	var out *os.File
	if *outfile == "-" {
		out = os.Stdout
	} else {
		var err error
		out, err = os.OpenFile(*outfile, os.O_CREATE|os.O_WRONLY, 0666)
		if err != nil {
			fmt.Println(err)
			os.Exit(exit_failure)
		}
		defer out.Close()
	}

	retval := exit_success

	if flag.NArg() <= 0 {
		doEscape(os.Stdin, out)
	} else {
		for i := 0; i < flag.NArg(); i++ {
			err := escapeFile(flag.Arg(i), out)
			if err != nil {
				fmt.Println(err)
				retval = exit_failure
			}
		}
	}

	os.Exit(retval)
}

File: c/escape.c
/* ********************************************************************** */
/**
 * @brief   Command line tool to escape characters greater than 0x7F.
 * @author  eel3
 * @date    2020/06/21
 *
 * @par Compilers
 * - TDM-GCC 4.8.1 (Windows 7 64bit SP1)
 * - Visual Studio 2005 SP1 (Windows 7 64bit SP1)
 * - Visual Studio 2010 SP1 (Windows 7 64bit SP1)
 */
/* ********************************************************************** */


#if defined(_WIN32) || defined(_WIN64)
#	if !defined(__MINGW32__) && !defined(__MINGW64__)
#		ifndef _CRT_SECURE_NO_WARNINGS
#			define _CRT_SECURE_NO_WARNINGS
#		endif /* ndef _CRT_SECURE_NO_WARNINGS */
#	endif /* !defined(__MINGW32__) && !defined(__MINGW64__) */
#endif /* defined(_WIN32) || defined(_WIN64) */

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


/* ---------------------------------------------------------------------- */
/* Constant */
/* ---------------------------------------------------------------------- */

/** File separator */
#if defined(_WIN32) || defined(_WIN64)
#	define  PATH_SEP  '\\'
#else /* defined(_WIN32) || defined(_WIN64) */
#	define  PATH_SEP  '/'
#endif /* defined(_WIN32) || defined(_WIN64) */


/* ---------------------------------------------------------------------- */
/* Variable */
/* ---------------------------------------------------------------------- */

/** The program name. */
static const char *program_name;


/* ---------------------------------------------------------------------- */
/* Function-like macro */
/* ---------------------------------------------------------------------- */

/* ====================================================================== */
/**
 * @brief  Return true if string s1 equal to s2.
 *
 * @param[in] s1  string 1.
 * @param[in] s2  string 2.
 *
 * @retval !=0  Equal.
 * @retval   0  Not equal.
 */
/* ====================================================================== */
#define  STREQ(s1, s2)  (((s1)[0] == (s2)[0]) && (strcmp((s1), (s2)) == 0))


/* ---------------------------------------------------------------------- */
/* Function */
/* ---------------------------------------------------------------------- */

/* ====================================================================== */
/**
 * @brief  Tiny copy of basename(3).
 *
 * @param[in] *name  (See man 3 basename).
 *
 * @return  (See man 3 basename).
 */
/* ====================================================================== */
static const char *my_basename(const char * const name)
{
	const char *bn;

	assert(name != NULL);

	bn = strrchr(name, PATH_SEP);
	return (bn == NULL) ? name : bn+1;
}

/* ====================================================================== */
/**
 * @brief  Show usage,
 *
 * @param[in,out] *out  Output stream.
 */
/* ====================================================================== */
static void usage(FILE * const out)
{
	assert(out != NULL);

	(void) fprintf(out, "usage: %s [-h] [file ...]\n", program_name);
}

/* ====================================================================== */
/**
 * @brief  Escape characters greater than 0x7F.
 *
 * @param[in,out] *in   input stream
 * @param[in,out] *out  output stream
 */
/* ====================================================================== */
static void do_escape(FILE *in, FILE *out)
{
	int c;

	assert((in != NULL) && (out != NULL));

	while ((c = fgetc(in)) != EOF) {
		if (c >= 0x80) {
			(void) fprintf(out, "\\x%02X", (unsigned) c);
		} else {
			(void) fputc(c, out);
		}
	}
}

/* ********************************************************************** */
/**
 * @brief  Main routine.
 *
 * @retval EXIT_SUCCESS  OK (success).
 * @retval EXIT_FAILURE  NG (failure).
 */
/* ********************************************************************** */
int main(int argc, char *argv[])
{
	int retval;

	program_name = my_basename(argv[0]);

	for (; (argc > 1) && (argv[1][0] == '-') && (argv[1][1] != '\0'); argc--, argv++) {
		const char *p = &argv[1][1];

		if (argv[1][1] == '-') {
			p = &argv[1][2];

			if (*p == '\0') {
				argc--, argv++;
				break;
			} else if (STREQ(p, "help")) {
				usage(stdout);
				return EXIT_SUCCESS;
			} else {
				usage(stderr);
				return EXIT_FAILURE;
			}
			continue;
		}

		do switch (*p) {
		case 'h':
			usage(stdout);
			return EXIT_SUCCESS;
		default:
			usage(stderr);
			return EXIT_FAILURE;
		} while (*++p != '\0');
	}

	retval = EXIT_SUCCESS;

	if (argc <= 1) {
		do_escape(stdin, stdout);
	} else {
		int i;
		FILE *in;

		for (i = 1; i < argc; ++i) {
			if (STREQ(argv[i], "-")) {
				in = stdin;
			} else if (errno = 0, (in = fopen(argv[i], "r")) == NULL) {
				perror(argv[i]);
				retval = EXIT_FAILURE;
				continue;
			} else {
				/*EMPTY*/
			}

			do_escape(in, stdout);

			if (in != stdin) {
				(void) fclose(in);
			}
		}
	}

	return retval;
}
